### ISSUE TITLE
Long running server error

### ISSUE THREAD

**Issue by AlbertDeFusco** [dask/dask-gateway#167]
I left my server running for several weeks and found this error. I'll leave it on again and check back.

```
2019/11/24 17:14:33 http: panic serving 202.107.226.5:63593: runtime error: index out of range
goroutine 358 [running]:
net/http.(*conn).serve.func1(0xc0001a9900)
	/opt/dask-gateway/miniconda/go/src/net/http/server.go:1769 +0x139
panic(0x6cc8e0, 0x99b5f0)
	/opt/dask-gateway/miniconda/go/src/runtime/panic.go:522 +0x1b5
main.normalizePath(...)
	/tmp/pip-req-build-9wxkzs12/dask-gateway-proxy/router.go:14
main.(*Router).HasMatch(0xc00001ec60, 0x0, 0x0, 0xc000073d10)
	/tmp/pip-req-build-9wxkzs12/dask-gateway-proxy/router.go:52 +0x1a0
main.(*HttpProxy).ServeHTTP(0xc000050500, 0x78ee80, 0xc0001d28c0, 0xc0001d5d00)
	/tmp/pip-req-build-9wxkzs12/dask-gateway-proxy/webproxy.go:96 +0x4d
net/http.serverHandler.ServeHTTP(0xc000066d00, 0x78ee80, 0xc0001d28c0, 0xc0001d5d00)
	/opt/dask-gateway/miniconda/go/src/net/http/server.go:2774 +0xa8
net/http.(*conn).serve(0xc0001a9900, 0x78f4c0, 0xc0001e6080)
	/opt/dask-gateway/miniconda/go/src/net/http/server.go:1878 +0x851
created by net/http.(*Server).Serve
	/opt/dask-gateway/miniconda/go/src/net/http/server.go:2884 +0x2f4
```

**Comment by jcrist** [comment]
Do you have any context around this? I can see where the error is, but not how we reached that state. Did this break anything else, or did you just see a traceback and things kept running as usual?

**Comment by AlbertDeFusco** [comment]
This is version 0.5.0

It kept running but I couldn't start a new cluster. It would throw an error about the `.crt` file not being present in `/tmp/dask-gateway-`.

here's the certificate error

```
[W 2019-11-25 20:04:30.092 DaskGateway] 401 POST /api/clusters/ (127.0.0.1) 1.95ms
[I 2019-11-25 20:04:30.268 DaskGateway] 201 POST /api/clusters/ (127.0.0.1) 167.01ms
[I 2019-11-25 20:04:30.268 DaskGateway] Starting cluster cc033bb1382e4014b6c7713f7d1c9bd4 for user instructor...
[E 2019-11-25 20:04:30.269 DaskGateway] Error while starting cluster cc033bb1382e4014b6c7713f7d1c9bd4
    Traceback (most recent call last):
      File "/opt/dask-gateway/miniconda/lib/python3.7/site-packages/dask_gateway_server/app.py", line 791, in start_cluster
        await self._start_cluster(cluster)
      File "/opt/dask-gateway/miniconda/lib/python3.7/site-packages/dask_gateway_server/app.py", line 777, in _start_cluster
        async for state in cluster.manager.start_cluster():
      File "/opt/dask-gateway/miniconda/lib/python3.7/site-packages/dask_gateway_server/managers/yarn.py", line 290, in start_cluster
        with self.temp_write_credentials() as (cert_path, key_path):
      File "/opt/dask-gateway/miniconda/lib/python3.7/contextlib.py", line 112, in __enter__
        return next(self.gen)
      File "/opt/dask-gateway/miniconda/lib/python3.7/site-packages/dask_gateway_server/managers/yarn.py", line 214, in temp_write_credentials
        with os.fdopen(os.open(path, flags, 0o600), "wb") as fil:
    FileNotFoundError: [Errno 2] No such file or directory: '/tmp/dask-gateway-yz3urc1f/cc033bb1382e4014b6c7713f7d1c9bd4.crt'
```

Found a few more errors related to TLS

```
[W 2019-11-23 08:04:31.130 SchedulerProxy] Extracting SNI: TLS record is not a handshake
[W 2019-11-23 08:04:34.298 SchedulerProxy] Extracting SNI: Error reading TLS record header: read tcp 172.31.4.108:8786->163.172.105.97:56260: i/o timeout
```

**Comment by jcrist** [comment]
Do you have any logs surrounding the error in the first post?

> Found a few more errors related to TLS

Unless these were caused by actions you know you took, these look like other clients trying (and failing) to connect, and are probably innocuous.

The error about missing certificates is odd, can you check and see if `/tmp/dask-gateway-yz3urc1f/` doesn't exist? This may be due to linux cleaning up a temporary directory that we were still using.

**Comment by jcrist** [comment]
> panic serving 202.107.226.5:63593

IP geolocation shows this request coming from china. I was pretty sure we sanitized all inputs, but clearly something is slipping through somewhere.

**Comment by AlbertDeFusco** [comment]
hmmm, I'll have to track that down. 

The /tmp/dask-gateway- directory was not present when I logged into the system today at 20:04:30.269 UTC, but the dask-gateway-server was still accessible to do anything but create a cluster.

**Comment by jcrist** [comment]
Ok, I found where we were missing sanitizing inputs, will have a PR fixing the top-level issue shortly.

I also know why the `/tmp` directory issue occurred (linux cleaned up the temporary directory we were using since it sat idle for a long time). Will have to decide how to solve this, but we know why for sure.

The only other question is if the TLS issues in:

> [W 2019-11-23 08:04:31.130 SchedulerProxy] Extracting SNI: TLS record is not a handshake
[W 2019-11-23 08:04:34.298 SchedulerProxy] Extracting SNI: Error reading TLS record header: read tcp 172.31.4.108:8786->163.172.105.97:56260: i/o timeout

were caused by a user action, or some external bot trying (and failing) to connect to our service. If they weren't caused by you/a user, then these are safe to ignore.
